### PR TITLE
chunked cross entropy loss

### PR DESCRIPTION
**Pull request by winglian**
in testing with qwen3, reduced single gpu vram from 22-26GB -> 15-18GB

## Summary by CodeRabbit

- **New Features**
  - Introduced an option to enable memory-efficient chunked cross-entropy loss during training, with configurable chunk sizes.
  - Added new configuration fields to control chunked cross-entropy behavior.
- **Bug Fixes**
  - Improved import handling for plugin integration.
- **Tests**
  - Added tests to verify the correctness of the chunked cross-entropy loss implementation.

### CODE REVIEW

File: src/axolotl/integrations/cut_cross_entropy/__init__.py [hunk 1]
```diff
@@ -72,7 +72,7 @@ def pre_model_load(self, cfg):
         if cfg.cut_cross_entropy:
             self._check_requirements()
 
-            from axolotl.integrations.cut_cross_entropy.monkeypatch.patch import (
+            from .monkeypatch.patch import (
```
Comment (NanoCode012) [inline src/axolotl/integrations/cut_cross_entropy/__init__.py:74]: Why are we changing to relative import here?

File: src/axolotl/monkeypatch/loss/chunked.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+chunked ce loss
+"""
+
+from typing import List, Optional
+
+import torch
+import torch.nn.functional as F
+
+
+# copied and modified from torchtune.modules.loss.CEWithChunkedOutputLoss
```
Comment (NanoCode012) [inline src/axolotl/monkeypatch/loss/chunked.py:11]: Do we need to add license block?

File: tests/test_chunked_xentropy.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    batch_size = 1
+
+    lm_head = nn.Linear(model_dim, vocab_size)
+    hidden_state = torch.randn(batch_size, seq_len, model_dim)
+    labels = torch.randint(low=0, high=vocab_size, size=(batch_size, seq_len))
+    return lm_head, hidden_state, labels, vocab_size
+
+
+def test_chunked_forward(chunked_fixtures):  # pylint: disable=redefined-outer-name
+    lm_head, hidden_state, labels, vocab_size = chunked_fixtures
```
Comment (NanoCode012) [inline tests/test_chunked_xentropy.py:26]: Should we set seed for RNG? (referring to torchtune's 

```py
from torchtune.utils.seed import set_seed


 @pytest.fixture(autouse=True)
 def random():
     set_seed(42)
```

File: src/axolotl/utils/schemas/config.py [hunk 4]
```diff
@@ -242,6 +242,9 @@ class AxolotlInputConfig(
     unsloth_rms_norm: bool | None = None
     unsloth_rope: bool | None = None
 
+    chunked_cross_entropy: bool | None = None
+    chunked_cross_entropy_num_chunks: int | None = None
+
```
Comment (NanoCode012) [inline src/axolotl/utils/schemas/config.py]: Can this be added to docs `config.qmd`?

File: src/axolotl/monkeypatch/loss/chunked.py (hunk 2 above)
Comment (NanoCode012) [reply src/axolotl/monkeypatch/loss/chunked.py:11]: The source is BSD3

File: src/axolotl/monkeypatch/loss/chunked.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        ]
+
+        # compute one chunk at a time
+        total_loss = 0.0
+        for logits_chunk, labels_chunk in zip(logits, labels):
+            total_loss += self.compute_cross_entropy(logits_chunk, labels_chunk)
+
+        if reduction == "sum":
+            return total_loss
+        return total_loss / total_elements
```
Comment (NanoCode012) [inline src/axolotl/monkeypatch/loss/chunked.py:71]: Should check if `mean` for second return and raise value/notimplemented error for other reduction

File: src/axolotl/monkeypatch/loss/chunked.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    ) -> torch.Tensor:
+        """
+        Upcast logits to fp32 and compute cross entropy loss.
+        """
+        return F.cross_entropy(
+            logits.float(), labels, ignore_index=self.ignore_index, reduction="sum"
+        )
+
+    def forward(
+        self, logits: List[torch.Tensor], labels: torch.Tensor, reduction="sum"
```
Comment (NanoCode012) [inline src/axolotl/monkeypatch/loss/chunked.py:38]:
```suggestion
        self, logits: List[torch.Tensor], labels: torch.Tensor, reduction="mean"
```

This should be mean to follow regular defaults? https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html#torch.nn.CrossEntropyLoss

File: src/axolotl/monkeypatch/loss/chunked.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        if reduction == "sum":
+            return total_loss
+        return total_loss / total_elements
+
+
+def _build_chunked_ce_loss_fn(num_output_chunks: int = 8, ignore_index: int = -100):
+    loss_fn_ce = CEWithChunkedOutputLoss(num_output_chunks, ignore_index)
+    loss_fn_ce.compute_cross_entropy = torch.compile(
+        loss_fn_ce.compute_cross_entropy, backend="inductor"
+    )
```
Comment (NanoCode012) [inline src/axolotl/monkeypatch/loss/chunked.py:78]: Should we also read env var like in torchtune?

```python
backend = os.environ.get("TORCH_COMPILE_BACKEND", "inductor")
```

File: src/axolotl/monkeypatch/loss/chunked.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    loss_fn_ce = _build_chunked_ce_loss_fn(num_output_chunks, ignore_index)
+
+    def chunked_fix_cross_entropy(
+        source,
+        target,
+        num_items_in_batch: int = None,
+        ignore_index: int = -100,
+        **kwargs,
+    ):  # pylint: disable=unused-argument
+        reduction = "sum" if num_items_in_batch is not None else "mean"
```
Comment (NanoCode012) [inline src/axolotl/monkeypatch/loss/chunked.py:92]: Is this correct? Could you explain the relationship between `num_items_in_batch` and `reduction`?

File: src/axolotl/monkeypatch/loss/chunked.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+        logit_chunks = [  # pylint: disable=unnecessary-comprehension
+            chunk for chunk in source.chunk(loss_fn_ce.num_output_chunks, dim=1)
```
Comment (NanoCode012) [inline src/axolotl/monkeypatch/loss/chunked.py:94]: Is this doing nothing, as current transformers already calculate the logits before hand?

Comment (NanoCode012) [reply src/axolotl/monkeypatch/loss/chunked.py:94]: I saw this comment on the linked PR

> It is important to notice that chunking must happen BEFORE the linear output, or the memory is not released:

Does it refer to the chunk CE inside the class or the logits chunk calculation?

File: src/axolotl/monkeypatch/loss/chunked.py (hunk 8 above)
Comment (NanoCode012) [reply src/axolotl/monkeypatch/loss/chunked.py:92]: Does that mean, we need to add a test with `num_items_in_batch` for `reduction=sum`?

**Review by NanoCode012 — COMMENTED** [review]
Nice implementation. Left some comments on parts I'm unsure of

File: src/axolotl/monkeypatch/loss/chunked.py (hunk 5 above)
Comment (NanoCode012) [reply src/axolotl/monkeypatch/loss/chunked.py:71]: Same for in the `get_*` function below.

File: src/axolotl/monkeypatch/loss/chunked.py (hunk 2 above)
Comment (salmanmohammadi) [inline src/axolotl/monkeypatch/loss/chunked.py:11]: This implementation has since been deprecated for https://github.com/pytorch/torchtune/blob/main/torchtune/modules/loss/cross_entropy_loss.py which we should follow instead

Comment (winglian) [reply src/axolotl/monkeypatch/loss/chunked.py:11]: Let's tackle that as a follow-on. The newer version in TT fuses with the linear by passing the weight of the embedding, but this isn't as straightforward and requires individual patches for each model architecture. This is a simpler patch that will work for all models

File: src/axolotl/monkeypatch/loss/chunked.py (hunk 6 above)
Comment (salmanmohammadi) [reply src/axolotl/monkeypatch/loss/chunked.py:38]: Do we not want to mean across all the chunks, rather than per-chunk? The actual normalisation happens below

File: src/axolotl/integrations/cut_cross_entropy/__init__.py (hunk 1 above)
Comment (winglian) [reply src/axolotl/integrations/cut_cross_entropy/__init__.py:74]: absolutely import would break in colab for some reason. Also, it's the pattern we're typically using in plugins.

File: tests/test_chunked_xentropy.py (hunk 3 above)
Comment (winglian) [reply tests/test_chunked_xentropy.py:26]: not really necessary?

File: src/axolotl/monkeypatch/loss/chunked.py (hunk 8 above)
Comment (winglian) [reply src/axolotl/monkeypatch/loss/chunked.py:92]: this is the stupid gradient accumulation think that Daniel Han brought up that forced this pattern.

File: src/axolotl/monkeypatch/loss/chunked.py (hunk 9 above)
Comment (winglian) [reply src/axolotl/monkeypatch/loss/chunked.py:94]: This is the chunking, the optimization is that by chunking before we upcast to float32, we reduce the VRAM overhead.

File: src/axolotl/monkeypatch/loss/chunked.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+            total_loss += self.compute_cross_entropy(logits_chunk, labels_chunk)
+
+        if reduction == "sum":
+            return total_loss
+        return total_loss / total_elements
+
+
+def _build_chunked_ce_loss_fn(num_output_chunks: int = 8, ignore_index: int = -100):
+    loss_fn_ce = CEWithChunkedOutputLoss(num_output_chunks, ignore_index)
+    loss_fn_ce.compute_cross_entropy = torch.compile(
```
Comment (NanoCode012) [inline src/axolotl/monkeypatch/loss/chunked.py:76]: Another question, would this work for all hardware? Do we need to check for compatibility?

Comment (salmanmohammadi) [reply src/axolotl/monkeypatch/loss/chunked.py:76]: Yeah unlike flex attention there's no cudagraphs/autotuning here so inductor should be pretty hardware friendly

**Review by salmanmohammadi — APPROVED** [review]
I think this looks good - the loss/grad norms lined up OK with the run you tested on?

Let's also make sure we add the license/code attribution link

**Comment by NanoCode012** [comment]
Rebased. Added license, arg to docs, and use TORCH_COMPILE_BACKEND for backend if available
